ctapipe/image/extractor.py

```diff
@@ -1285,3 +1289,406 @@ def __call__(
+        Gain loss of each channel that needs to be corrected after deconvolution.
+    time_shifts_nsec : list of floats
+        Timing shift of each channel that needs to be corrected after deconvolution.
+    leading_edge_shifts : list of floats
+        Offset of the leading edge peak w.r.t. the deconvolved peak.
+    """
+    if upsampling < 1:
+        raise ValueError(f"upsampling must be > 0, got {upsampling}")
+    if window_width < 1:
+        raise ValueError(f"window_width must be > 0, got {window_width}")
+
+    ref_pulse_shapes = camera.readout.reference_pulse_shape
+    ref_sample_width_nsec = camera.readout.reference_pulse_sample_width.to_value(u.ns)
+    camera_sample_width_nsec = 1.0 / camera.readout.sampling_rate.to_value(u.GHz)
+
+    if camera_sample_width_nsec < ref_sample_width_nsec:
+        raise ValueError(
+            f"Reference pulse sampling time (got {ref_sample_width_nsec} ns) must be equal to or shorter than the "
+            f"camera sampling time (got {camera_sample_width_nsec} ns); need a reference single p.e. pulse shape with "
+            "finer sampling!"
+        )
+    avg_step = int(round(camera_sample_width_nsec / ref_sample_width_nsec))
+
+    pole_zeros = []  # avg. pole-zero deconvolution parameters
+    for ref_pulse_shape in ref_pulse_shapes:
+        phase_pzs = []
+        for phase in range(avg_step):
+            x = ref_pulse_shape[phase::avg_step]
+            i_min = np.argmin(np.diff(x)) + 1
+            phase_pzs.append(x[i_min + 1] / x[i_min])
+
+        if len(phase_pzs) == 0:
+            raise ValueError(
+                "ref_pulse_shape is malformed - cannot find deconvolution parameter"
+            )
+        pole_zeros.append(np.mean(phase_pzs))
+
+    gains, shifts, pz2d_shifts = [], [], []  # avg. gains and timing shifts
+    for pz, ref_pulse_shape in zip(pole_zeros, ref_pulse_shapes):
+        if time_profile_pdf:  # convolve ref_pulse_shape with time profile PDF
+            t = (
+                np.arange(ref_pulse_shape.size) - ref_pulse_shape.size / 2
+            ) * ref_sample_width_nsec
+            time_profile = time_profile_pdf(t)
+            ref_pulse_shape = np.convolve(ref_pulse_shape, time_profile, "same")
+
+        integral = (
+            ref_pulse_shape.sum() * ref_sample_width_nsec / camera_sample_width_nsec
+        )
+        phase_gains, phase_shifts, phase_pz2d_shifts = [], [], []
+        for phase in range(avg_step):
+            x = np.atleast_2d(ref_pulse_shape[phase::avg_step])
+            y = deconvolve(x, 0.0, upsampling, pz)[0]
+            i_max = y.argmax()
+            start = i_max - window_shift
+            stop = start + window_width
+            if start >= 0 and stop <= y.size:
+                if leading_edge_timing:
+                    d = deconvolve(x, 0.0, upsampling, 1)[0]
+                    d_pk_idx = d.argmax()
+                    phase_pz2d_shifts.append(i_max - d_pk_idx)
+                    time = adaptive_centroid(
```

Comment:
`adaptive_centroid` seems to take 4 arguments, but I can only see that 3 are provided, is this intentional? (Found by codacy)

Reply:
This is how the `guvectorize` decorator works, it changes the signature and the missing argument is actually the return value, that gets pre-allocated by the decorator, passed to the compiled function as mutable input argument but then returned. 

https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-guvectorize-decorator

Reply:
Unfortunately, this is somewhat opaque to static analysis